src/frequenz/sdk/timeseries/_ringbuffer.py

```diff
@@ -0,0 +1,431 @@
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from datetime import datetime, timedelta
+from typing import Any, Dict, Generic, List, Sequence, TypeVar
+
+import numpy as np
+from attr import dataclass
+
+T = TypeVar("T")
+
+Container = TypeVar("Container", list, np.ndarray)
+
+
+@dataclass
+class Gap:
+    """A gap defines the range for which we haven't received values yet."""
+
+    start: datetime
+    """Start of the range, inclusive."""
+    end: datetime
+    """End of the range, exclusive."""
+
+
+class OrderedRingBuffer(Generic[T, Container]):
+    """Time aware ringbuffer that keeps its entries sorted by time."""
+
+    def __init__(
+        self,
+        buffer: Container,
+        sampling_period: timedelta,
+        time_index_alignment: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """Initialize the time aware ringbuffer.
+
+        Args:
+            buffer: Instance of a buffer container to use internally.
+            sampling_period: Timedelta of the desired resampling period.
+            time_index_alignment: Arbitary point in time used to align
+                timestamped data with the index position in the buffer.
+                Used to make the data stored in the buffer align with the
+                beginning and end of the buffer borders.
+
+                For example, if the `time_index_alignment` is set to
+                "0001-01-01 12:00:00", and the `sampling_period` is set to
+                1 hour and the length of the buffer is 24, then the data
+                stored in the buffer could correspond to the time range from
+                "2022-01-01 12:00:00" to "2022-01-02 12:00:00" (date chosen
+                arbitrarily here).
+        """
+        self._buffer = buffer
+        self._sampling_period = sampling_period
+        self._time_index_alignment = time_index_alignment
+
+        self._gaps: list[Gap] = []
+        self._datetime_newest = datetime.min
+        self._datetime_oldest = datetime.max
+        self._time_range = (len(self._buffer) - 1) * sampling_period
+
+    @property
+    def sampling_period(self) -> timedelta:
+        """Return the sampling period of the ring buffer.
+
+        Returns:
+            Sampling period of the ring buffer.
+        """
+        return self._sampling_period
+
+    @property
+    def gaps(self) -> List[Gap]:
+        """Get the list of ranges for which no values were provided.
+
+        See definition of dataclass @Gaps for more info.
+
+        Returns:
+            List of gaps.
+        """
+        return self._gaps
+
+    @property
+    def maxlen(self) -> int:
+        """Get the max length.
+
+        Returns:
+            The max amount of items this container can hold.
+        """
+        return len(self._buffer)
+
+    def update(self, timestamp: datetime, value: T, missing: bool = False) -> bool:
```

Comment:
This function raises in `TypeError` inherited from the `datetime_to_index` method.

Reply:
Any way to reproduce that?

Reply:
It raises when the other method does. I meant it should be documented :)

Reply:
Actually no, it's correct like this. It can't raise that `TypeError` because the conditions checked in `datetime_to_index` are modified in update so that it would never through